src/engine/enginebufferscalelinear.cpp

```diff
@@ -123,20 +120,64 @@ CSAMPLE* EngineBufferScaleLinear::getScaled(unsigned long buf_size) {
         m_dOldRate = 0.0;
         m_dRate = rate_add_new;
         // pass the address of the sample at the halfway point
-        do_scale(&m_buffer[buf_size / 2], buf_size / 2, &samples_read);
+        samples_read += do_scale(&pOutput[buf_size / 2], buf_size / 2);
+        return samples_read;
+    }
+
+    samples_read += do_scale(pOutput, buf_size);
+    return samples_read;
+}
 
-        m_samplesRead = samples_read;
-        return m_buffer;
+int EngineBufferScaleLinear::do_copy(CSAMPLE* buf, const int buf_size) {
+    int samples_needed = buf_size;
+    CSAMPLE* write_buf = buf;
+    // Use up what's left of the internal buffer.
+    int iNextSample = static_cast<int>(ceil(m_dNextFrame)) * 2;
+    int readSize = math_min<int>(m_bufferIntSize - iNextSample, samples_needed);
+    if (readSize > 0) {
+        SampleUtil::copy(write_buf, &m_bufferInt[iNextSample], readSize);
+        samples_needed -= readSize;
+        write_buf += readSize;
+    }
+    // Protection against infinite read loops when (for example) we are
+    // reading from a broken file.
+    int read_failed_count = 0;
+    // We need to repeatedly call the RAMAN because the RAMAN does not bend
+    // over backwards to satisfy our request. It assumes you will continue
+    // to call getNextSamples until you receive the number of samples you
+    // wanted.
+    while (samples_needed > 0) {
+        int read_size = m_pReadAheadManager->getNextSamples(m_dRate, write_buf,
+                samples_needed);
+        if (read_size == 0) {
+            if(++read_failed_count > 1) {
+                break;
+            } else {
+                continue;
+            }
+        }
+        samples_needed -= read_size;
+        write_buf += read_size;
     }
 
-    CSAMPLE* result = do_scale(m_buffer, buf_size, &samples_read);
-    m_samplesRead = samples_read;
-    return result;
+    // Instead of counting how many samples we got from the internal buffer
+    // and the RAMAN calls, just measure the difference between what we
+    // requested and what we still need.
+    int read_samples = buf_size - samples_needed;
+    // Zero the remaining samples if we didn't fill them.
+    SampleUtil::clear(write_buf, samples_needed);
+    // update our class members so next time we need to scale it's ok. we do
+    // blow away the fractional sample position here
+    m_bufferIntSize = 0; // force buffer read
+    m_dNextFrame = 0;
+    m_floorSampleOld[0] = buf[read_samples - 2];
```

Comment:
There are more suspect index calculations in this code that should at least be protected by a DEBUG_ASSERT and maybe a comment why this is safe.